plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -2040,22 +2040,37 @@ private ThriftMetastoreClient createMetastoreClient(HiveIdentity identity)
 
         String username = identity.getUsername().orElseThrow(() -> new IllegalStateException("End-user name should exist when metastore impersonation is enabled"));
         if (authenticationEnabled) {
-            String delegationToken;
             try {
-                delegationToken = delegationTokenCache.getUnchecked(username);
+                return createMetastoreClientWithAuthentication(username, false);
             }
-            catch (UncheckedExecutionException e) {
-                throwIfInstanceOf(e.getCause(), TrinoException.class);
-                throw e;
+            catch (TException e) {
+                log.debug(e, "It may be a delegation token permission failure, so try again with a new delegation token");
+                return createMetastoreClientWithAuthentication(username, true);
             }
-            return clientProvider.createMetastoreClient(Optional.of(delegationToken));
         }
 
         ThriftMetastoreClient client = createMetastoreClient();
         setMetastoreUserOrClose(client, username);
         return client;
     }
 
+    private ThriftMetastoreClient createMetastoreClientWithAuthentication(String username, boolean refresh)
```

Comment:
Extract:

```java
 private String getDelegationToken(String username)
    {
        String delegationToken;
        try {
            delegationToken = delegationTokenCache.getUnchecked(username);
        }
        catch (UncheckedExecutionException e) {
            throwIfInstanceOf(e.getCause(), TrinoException.class);
            throw e;
        }
        return delegationToken;
    }
```

and inline this back:
```java
        if (authenticationEnabled) {
            try {
                return clientProvider.createMetastoreClient(Optional.of(getDelegationToken(username)));
            }
            catch (TException e) {
                log.debug(e, "It may be a delegation token permission failure, so try again with a new delegation token");
                delegationTokenCache.refresh(username);
                return clientProvider.createMetastoreClient(Optional.of(getDelegationToken(username)));
            }
        }
```

Reply:
modified in that way. please check again.